runtime/doc/lua-guide.txt

```diff
@@ -0,0 +1,450 @@
+	vim.keymap.set('n', '<Leader>ex2', function() print("Example 2") end)
+	vim.keymap.set('n', '<Leader>pl1', require('plugin').plugin_action)
+	-- To avoid the startup cost of requiring the module, you can wrap it in a function to require it lazily when invoking the mapping:
+	vim.keymap.set('n', '<Leader>pl2', function() require('plugin').plugin_action() end)
+
+The fourth (optional) argument is a table containing boolean options for the
+mapping as defined in |:map-arguments|, and also `replace_keycodes`,
+`remap` and `desc`. >lua
+
+	vim.keymap.set('n', '<Leader>ex1', '<Cmd>echomsg "Example 1"<CR>', {buffer = true})
+	vim.keymap.set('n', '<Leader>ex2', function() print('Example 2') end, {desc = 'Prints "Example 2" to the message area'})
+
+Defining mappings with a Lua function is different from using a string. The
+usual way to show information about a keymap like `:nmap <Leader>ex1` will not
+output useful information (the string itself), but instead only show "Lua
+function". It is recommended to add a `desc` key to describe the behavior of
+your mapping. This is especially important for documenting plugin mappings so
+users can understand the usage of the mapping more easily.
+
+Mappings are `noremap` by default: >lua
+
+    vim.keymap.set('n', '<Leader>test1', '<Cmd>echo "test"<CR>')
+    -- :nnoremap <Leader>test <Cmd>echo "test"<CR>
+
+    -- If you DO want the mapping to be recursive, set the "remap" option to "true"
+    vim.keymap.set('n', '>', ']', {remap = true})
+    -- :nmap > ]
+
+-------------------------------------------------------------------------------
+vim.keymap.del()
+
+|vim.keymap.del()| is a convenience function of API functions
+vim.api.|nvim_del_keymap()| and vim.api.|nvim_buf_del_keymap()|
+
+|vim.keymap.del()| works the same way as |vim.keymap.set()| but deletes
+mappings instead: >lua
+
+	vim.keymap.del('n', '<Leader>ex1')
+	vim.keymap.del({'n', 'c'}, '<Leader>ex2', {buffer = true})
+
+
+See also:
+• vim.api.|nvim_get_keymap()|     - list all global mapping
+• vim.api.|nvim_buf_get_keymap()| - list all mappings for buffer
+
+==============================================================================
+Defining autocommands
+TODO: write
+
+Looks useful:
+• nvim_del_autocmd
+
+Looks decent:
+• nvim_clear_autocmds
+• nvim_get_autocmds
+
+Looks meh:
+• nvim_create_augroup
+• nvim_del_augroup_by_id
+• nvim_del_augroup_by_name
+
+wat:
+• nvim_exec_autocmds
+
+TODO: what's the point of manually executing an autocommand?
+
+-------------------------------------------------------------------------------
+Creating autocommands
+
+Use vim.api.|nvim_create_autocmd()| to create autocommands from Lua. Lua
+autocommands can execute a command (like regular autocommands), but they can
+also call a function. We call these "callback functions" since we don't
+immediately execute them, we only tell Neovim to run them some time in the
+future if the autocommand is triggered.
```

Comment:
I would explain autocommands as something like "running an action on some event", and then say that it can be an ex command or a lua function. Since this is a guide, if a mechanism can be briefly explained I think we should just do that, instead of assuming that the reader knows vim script and explaining in terms of how it relates to it.

Reply:
Hmm, I'm not a fan, sorry. I think assuming the reader has some previous knowledge of vim (and stating this at the intro) helps the guide be more focused. I'm not completely against the idea.

@clason thoughts?

Reply:
I agree that it should not start from `:Tutor` and assume some basics, but if you can remind people with a single sentence, that is always welcome.